### PR TITLE
Add ResponseBodyMatches method

### PR DESCRIPTION
**Pull request by irpap**

### CODE REVIEW

File: http.go [hunk 1]
```diff
@@ -270,6 +272,22 @@ func (hc *HttpCall) ResponseBodyContains(value string) Step {
 	})
 }
 
+// ResponseBodyMatches is a Step that when executed ensures there is
+// a non-nil hc.ResponseBody and errors unless the hc.ResponseBody
+// matches the regular expression parameter.
+func (hc *HttpCall) ResponseBodyMatches(pattern string) Step {
```
Comment (msackman) [inline http.go]: Thanks for the PR. In the interests of flexibility, I would prefer this to take a *Regexp rather than a string.

File: http.go [hunk 2]
```diff
@@ -270,6 +272,22 @@ func (hc *HttpCall) ResponseBodyContains(value string) Step {
...
+	return NewNamedStep("ResponseBodyMatches", func() error {
+		if err := hc.ReceiveBody(); err != nil {
+			return err
+		}
+		var regex = regexp.MustCompile(pattern)
```
Comment (msackman) [inline http.go]: I would also prefer that the whole body is a single if-elseif-else block in keeping with the vast majority of the other steps here.

File: http.go [hunk 3]
```diff
@@ -270,6 +272,22 @@ func (hc *HttpCall) ResponseBodyContains(value string) Step {
...
+		if !regex.MatchString(string(hc.ResponseBody)) {
+			return fmt.Errorf("Body: Expected to match the pattern '%s'; found '%s'.", pattern, string(hc.ResponseBody))
```
Comment (msackman) [inline http.go]: *Regexp does indeed have a sensible String() method, so this error makes a lot of sense. But that also means that in the first arg to NewNamedStep you can use fmt.Sprintf and have a more useful name - again in keeping with all the other steps.

**Review by msackman — CHANGES_REQUESTED** [review]

File: http.go [hunk 4]
```diff
@@ -275,13 +275,11 @@ func (hc *HttpCall) ResponseBodyContains(value string) Step {
 // ResponseBodyMatches is a Step that when executed ensures there is
 // a non-nil hc.ResponseBody and errors unless the hc.ResponseBody
 // matches the regular expression parameter.
-func (hc *HttpCall) ResponseBodyMatches(pattern string) Step {
-	return NewNamedStep("ResponseBodyMatches", func() error {
+func (hc *HttpCall) ResponseBodyMatches(pattern *regexp.Regexp) Step {
+	return NewNamedStep(fmt.Sprintf("ResponseBodyMatches(%s)", pattern), func() error {
```
Comment (msackman) [inline http.go]: I don't think you've tested this ;) `%s` is only for when the arg is a string. You want `%v` here, and below in your `Errorf` call.

File: http.go [hunk 5]
```diff
@@ -275,13 +275,11 @@ func (hc *HttpCall) ResponseBodyContains(value string) Step {
...
 		if err := hc.ReceiveBody(); err != nil {
 			return err
-		}
-		var regex = regexp.MustCompile(pattern)
-		if !regex.MatchString(string(hc.ResponseBody)) {
+		} else if !pattern.MatchString(string(hc.ResponseBody)) {
 			return fmt.Errorf("Body: Expected to match the pattern '%s'; found '%s'.", pattern, string(hc.ResponseBody))
 		}
 		return nil
```
Comment (msackman) [inline http.go]: I would still put this in an `else`, like the others. I just really like the symmetry of that - it's more functional ;)

**Review by msackman — CHANGES_REQUESTED** [review]

File: http.go (hunk 4 above)
Comment (irpap) [reply http.go]: I ran a test using it, i didn't see any output so I guess I didn't test it right.

Comment (msackman) [reply http.go]: Apologies - I've just done some digging and this does not error which surprises me - just been reading the docs on `fmt` and it's more subtle than I thought.

```
package main
import "fmt"

type Thing struct{}
func (t *Thing) String() string {
	return "weird"
}

func main() {
	fmt.Println(fmt.Sprintf("%s", &Thing{}))
}
```

I assumed the above would error as Thing is not a string. But it turns out there are special rules that allow this to work. I'm afraid I would still prefer `%v` though, but I apologise for suggesting you'd not tested it.

Comment (irpap) [reply http.go]: Np!

File: http.go (hunk 5 above)
Comment (msackman) [reply http.go]: And this one?

**Review by msackman — APPROVED** [review]

**Comment by msackman** [comment]
Perfection itself. Thank you :)

**Pull request merged**